tests/debugpy/test_run.py

```diff
@@ -137,44 +137,90 @@ def code_to_debug():
         assert backchannel.receive() == "1"
 
 
-@pytest.mark.parametrize("run", runners.all_launch)
-@pytest.mark.parametrize("python_args", [None, "-B"])
-@pytest.mark.parametrize("python", [None, "custompy", "custompy,-O"])
-@pytest.mark.parametrize("python_key", ["python", "pythonPath"])
-def test_custom_python(pyfile, tmpdir, run, target, python_key, python, python_args):
+def make_custompy(tmpdir, id=""):
     if sys.platform == "win32":
-        custompy = tmpdir / "custompy.cmd"
+        custompy = tmpdir / ("custompy" + id + ".cmd")
         script = """@echo off
-            set DEBUGPY_CUSTOM_PYTHON=1%DEBUGPY_CUSTOM_PYTHON%
+            set DEBUGPY_CUSTOM_PYTHON=<id>;%DEBUGPY_CUSTOM_PYTHON%
             <python> %*
             """
     else:
-        custompy = tmpdir / "custompy.sh"
+        custompy = tmpdir / ("custompy" + id + ".sh")
         script = """#!/bin/sh
-            env DEBUGPY_CUSTOM_PYTHON=1$DEBUGPY_CUSTOM_PYTHON <python> "$@"
+            env DEBUGPY_CUSTOM_PYTHON=<id>;$DEBUGPY_CUSTOM_PYTHON <python> "$@"
             """
-    custompy.write(script.replace("<python>", sys.executable))
+
+    script = script.replace("<id>", id)
+    script = script.replace("<python>", sys.executable)
+    custompy.write(script)
     os.chmod(custompy.strpath, 0o777)
 
+    return custompy.strpath
+
+
+@pytest.mark.parametrize("run", runners.all_launch)
+@pytest.mark.parametrize("debuggee_custompy", [None, "launcher"])
+@pytest.mark.parametrize("launcher_custompy", [None, "debuggee"])
+def test_custom_python(
+    pyfile, tmpdir, run, target, debuggee_custompy, launcher_custompy,
+):
     @pyfile
     def code_to_debug():
         import os
+
+        import debuggee
+        from debuggee import backchannel
+
+        debuggee.setup()
+        backchannel.send(os.getenv("DEBUGPY_CUSTOM_PYTHON"))
+
+    expected = ""
+    if debuggee_custompy:
+        debuggee_custompy = make_custompy(tmpdir, "debuggee")
+        expected += "debuggee;"
+    if launcher_custompy:
+        launcher_custompy = make_custompy(tmpdir, "launcher")
+        expected += "launcher;"
+    else:
+        # If "python" is set, it also becomes the default for "debugLauncherPython"
```

Comment:
I think a consequence of this is that the Docker extension has to pass in the Python interpreter path, which is not ideal. IMO if `debugLauncherPython` is unset, it should be the same as `debugAdapterPython`. In turn if that is unset, both should be same as the Python interpreter path. `python` should control only what the launcher uses, or otherwise I'd say add a new property so that only that one particular thing can be controlled.

Reply:
Er, I didn't notice this was test code; but I'm unsure--which of the three properties does Docker need to set?

Reply:
Keep in mind that this is going to work in conjunction with the Python extension in all VSCode scenarios. The extension sets all three properties to the current active interpreter for the workspace by default when resolving the configuration, skipping those that are already set. Thus, in your resolver, you would only need to set "python" to whatever it should be in the container, and let the extension provide the default for "debugLauncherPython". 

But this branch would kick in for other DAP clients, such as VS or Vimspector, and maintain existing behavior for them.

Reply:
Sounds good, thanks!